README.md
**OutlierDenStream**
===================


OutlierDenStream is a custom implementation of the [DenStream] algorithm used for Anomaly Detection in an Unsupervised environment written in Python.  The implementation has been developed for a project funded by NewNet@Paris, Cisco's Chair at Telecom ParisTech, in the context of the telemetry project https://telemetry.telecom-paristech.fr/


Description
-------------


The algorithm uses the concept of micro-cluster, distances, parameters and pruning strategies introduced in DenStream. It maintains the clusters incrementally and as soon as a new sample (snapshot) is available it labels it as Normal or Abnormal: e.g. if the new sample is merged to a existing core-micro-cluster then it is considered normal.

The algorithm can be used to monitor **Telemetry** and raise an alarm when needed. Each sample (data point) should have **1 x F** dimensionality where F is the number of features.


Documents
-------------


The first phase of the algorithm consist in discovering the clusters thus there is the need first of all to gather samples or use a buffer dataset then run then DBScan algorithm and obtain the clusters. From now on it is possible to maintain them incrementally. On the other hand it is possible to use the buffer dataset as a unique cluster without applying DBScan such that all the samples are considered as belonging to a cluster. 

The input parameters are the following one:

> * **lamb**: the fading factor <img src="https://latex.codecogs.com/gif.latex?\lambda" title="\lambda" />
> * **epsilon**: the radius <img src="https://latex.codecogs.com/gif.latex?\epsilon" title="\epsilon" />
>   * **'auto'**: computes automatically the radius of the initial cluster (if you don't use the initial DBScan)
>   * **int** or **float**: <img src="https://latex.codecogs.com/gif.latex?\epsilon" title="\epsilon" /> value integer of float
> * **minPts**: DBScan parameter, if used
> * **beta**: potential factor <img src="https://latex.codecogs.com/gif.latex?\beta" title="\beta" />
> * **mu**: cluster weight <img src="https://latex.codecogs.com/gif.latex?\mu" title="\mu" />
>   * **'auto'**: computes automatically the maximum weight of the cluster, due to fading function
>   * **int** or **float**:  <img src="https://latex.codecogs.com/gif.latex?\mu" title="\mu" /> value integer or float
> * **numberInitialSample**: number of samples needed before starting the DBScan algorithm on the gathered samples. Only if you use DBSCan.
> * **startingBuffer**: buffer with the initial samples
> * **tp**: checking period of the clusters weight. Needed for pruning, if the weight of the clusters goes below the threshold <img src="https://latex.codecogs.com/gif.latex?\beta&space;\cdot&space;\mu" title="\beta \cdot \mu" />: remove them.

Example:

```python
from DenStream import DenStream
...
den = DenStream (lamb=0.03, epsilon='auto', beta=0.03, mu='auto', startingBuffer=bufferDf, tp=12)

den.runInitialization()
```

This example shows the case in which a buffer is passed to the algorithm, thus there is not the need to wait for the incoming data. Moreover **epsilon** and **mu** are set to **'auto'** thus we are not going to perform the initial DBScan meaning that the samples passed in the buffer have to be considered as samples belonging to a unique initial cluster and the parameters can be computed by this last one. The method **.runInitialization()** performs indeed these tasks and in particular creates a cluster by the samples and computes <img src="https://latex.codecogs.com/gif.latex?\epsilon" title="\epsilon" /> and <img src="https://latex.codecogs.com/gif.latex?\mu" title="\mu" />.

The algorithm is now able to maintain the clusters incrementally. As soon as a new sample is available it is possible to call the **.onNewSample(Sample sample)** method. This one takes as input the object **Sample** and returns the label.

Example:
```python
from sample import Sample
...
for row in dataset:
    sample = Sample(row, timestamp)
    result = den.runOnNewSample(sample)
```

The example shows the simulation of an experiment in which we have a dataset <img src="https://latex.codecogs.com/gif.latex?X&space;\in&space;R&space;^&space;{N,&space;F}" title="X \in R ^ {N, F}" /> where N is the number of samples and F of features. Each row of F elements is extracted from the dataset and a **Sample** object is instantiated by it. The Sample object is composed by the main attributes **.value** and **.timestamp** which are respectively the array of the F measurements and the time (the timestamp is used just to retrieve the real timestamp, not for clustering purpose).
Thus on each new available sample, the **.runOnNewSample()** method is run and the *Normal* or *Abnormal* label is obtained.  

Installation
---

Clone the repository and import in your project DenStream and Sample

```python
from sample import Sample
from DenStream import DenStream
...
```

Related repositories
--------------------
The code released in this website are also instrumental to reproduce results that are published in [ACM SIGCOMM BigDama'18] and that are demonstrated at [IEEE INFOCOM'18] (see the Reference section below)

This repository only contains the algorithm, whereas related repositories contain
- the datasets we released to the community https://github.com/cisco-ie/telemetry/blob/master/README.md
- specific instruction and code to replicate the paper results https://github.com/anrputina/OutlierDenStream-BigDama18

Demo
---

A demo of the algorithm is available here: https://telemetry.telecom-paristech.fr/

Dataset
---

The Dataset in use is extracted from: https://github.com/cisco-ie/telemetry

References
---------

[ACM SIGCOMM BigDama'18] Putina, Andrian and Rossi, Dario and Bifet, Albert and Barth, Steven and Pletcher, Drew and Precup, Cristina and Nivaggioli, Patrice,  Telemetry-based stream-learning of BGP anomalies ACM SIGCOMM Workshop on Big Data Analytics and Machine Learning for Data Communication Networks (Big-DAMA’18) aug. 2018

[IEEE INFOCOM'18] Putina, Andrian and Rossi, Dario and Bifet, Albert and Barth, Steven and Pletcher, Drew and Precup, Cristina and Nivaggioli, Patrice,  Unsupervised real-time detection of BGP anomalies leveraging high-rate and fine-grained telemetry data IEEE INFOCOM, Demo Session apr. 2018,

[DenStream] Feng Cao, Martin Estert, Weining Qian, and Aoying Zhou, "Density-Based Clustering over an Evolving Data Stream with Noise" in Proceedings of the 2006 SIAM International Conference on Data Mining. 2006, 328-339 


HISTORY.rst
=======
History
=======

0.0.1 (2020-10-09)
------------------

* First release on PyPI.


Examples/ExampleClearBGPoldAll.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Mon Nov 13 17:19:09 2017

@author: anr.putina
"""

import sys
sys.path.append('./../')

import time
import numpy as np
import pandas as pd

from multiprocessing import Process

from sample import Sample
from DenStream import DenStream
from statistics import Statistics

from groundTruth import groundTruth

from visualization import Visualization
def normalize_matrix(df):
    return (df - df.mean())/df.std()

#simulationDataset = 'PortFlap'
simulationDataset = 'BGP_CLEAR'
simulationDataset = 'BGP_CLEAR2'

nodes = ['leaf1', 'leaf2', 'leaf3', 'leaf5', 'leaf6', 'leaf7', 'leaf8',\
         'spine1', 'spine2', 'spine3', 'spine4']

#nodes = ['leaf2']

def worker(beta):

    lambdas = np.arange(0.02, 0.03, 0.01)
    lambdas = [0.03]

    results = []
    for lamb in lambdas:

        resultSimulation = {}

        for simulationDataset in ['BGP_CLEAR']:
        
            for node in nodes:
        
                if simulationDataset == 'PortFlap':
                    
                    df = pd.read_csv('Data/'+node+'base_no_traffic.csv').dropna()\
                            .drop('Unnamed: 0', axis=1)\
#                            .drop('MgmtEth0/RP0/CPU0/0reliability', axis=1)\
#                            .drop('MgmtEth0/RP0/CPU0/0packets-sent', axis=1)
                            
                elif simulationDataset == 'BGP_CLEAR':
                    ### DATA PLANE
                    df = pd.read_csv('Data/'+node+'_clearbgp.csv').dropna()\
                            .drop('Unnamed: 0', axis=1)\
#                            .drop('MgmtEth0/RP0/CPU0/0packets-sent', axis=1)\
                #            .drop('HundredGigE0/0/0/30packets-sent', axis=1)   
                elif simulationDataset == 'BGP_CLEAR2':
                    ### DATA PLANE
                    df = pd.read_csv('Data/'+node+'bgp_clear_NoDataPlane.csv').dropna()\
                            .drop('Unnamed: 0', axis=1)\
        
                else:
                    print 'CHECK DATASET'
        
                df = df[:500]
        
                df['time'] = df['time'] / 1000000    
                times = df['time']
                df = df.drop(['time'], axis=1)
        
                featuresDrop = []
                for feature in df.columns:
                    if 'MgmtEth' in feature:
                        featuresDrop.append(feature)
                    if 'reliability' in feature:
                        featuresDrop.append(feature)
                    if 'Hundred' in feature:
                        featuresDrop.append(feature)
        
                df = df.drop(featuresDrop, axis=1)
        
                if 'cluster' in df:    
                    df = df.drop('cluster', axis=1)
        
        
                dfNormalized = normalize_matrix(df).dropna(axis=1)
        
                sampleSkip = 41
                bufferDf = dfNormalized[0:sampleSkip]
                testDf = dfNormalized[sampleSkip:]
        
                den = DenStream(lamb=lamb, epsilon='auto', beta=beta, mu='auto', startingBuffer=bufferDf, tp=12)
                den.runInitialization()
        
                outputNEW = []
                startingSimulation = time.time()
                for sampleNumber in range(len(testDf)):
                    sample = testDf.iloc[sampleNumber]
                    result = den.runOnNewSample(Sample(sample.values, times.iloc[sampleNumber]))
                    outputNEW.append(result)
                ### END SIMULATION ###
                print time.time() - startingSimulation
        
                df['result'] = [False] * sampleSkip + outputNEW
                
                truth = groundTruth()
                if simulationDataset == 'PortFlap':
                    truth.simulationPORT_FLAP()
                    truth.simulationPORT_FLAP_CLEAR()
                elif simulationDataset =='BGP_CLEAR':
                    truth.simulationBGP_CLEAR3_TwoMin()
                    truth.simulationBGP_CLEAR_CLEAR()
                elif simulationDataset =='BGP_CLEAR2':
                    truth.simulationBGP_CLEAR_Second_DATASET()
                    truth.simulationBGP_CLEAR2_CLEAR()
        
        #        visual=Visualization()
        #        visual.plotResults(df, times, truth)
        
                statistics = Statistics(node, truth)
                resultSimulation[node+simulationDataset] = statistics.getNodeResult(df, times, kMAX=5)

        statistics.getPrecisionRecallFalseRate(resultSimulation, kMAX=5, plot=True)
        statistics.getDelay(resultSimulation, kMAX=5, plot=True)
        
        #    result = {
        #        'PRF': statistics.getPrecisionRecallFalseRate(resultSimulation, kMAX=5, plot=False),
        #        'Delay': statistics.getDelay(resultSimulation, kMAX=5, plot=False)
        #    }
        
#        results.append(statistics.getPrecisionRecallFalseRate(resultSimulation, kMAX=5, plot=True))

    

#    filename = 'beta'+str(beta)
#    df = pd.DataFrame(results)
#    df.to_csv(filename+'.csv')
    
#    filename = 'beta'+str(beta)
#    import json
#    with open('Results/'+filename+'.txt', 'w') as outfile:
#       json.dump({beta:results}, outfile)
       
       
    return results

if __name__ == '__main__':
    
#    betas = np.arange(0.03, 0.01)
    worker(0.03)
#    for beta in betas:
#
#        p = Process(target=worker, args=(beta,))
#        p.start()
        


# #features = ['Precision, Recall']
# #Ks = [2, 3]
# #
# #arrays = np.zeros((len(features), len(Ks)))
# #
# Precision2 = []
# Precision3 = []

# Recall2 = []
# Recall3 = []
# for result in results:
    
#     Precision2.append(result['PRF']['Precision'][2])
#     Precision3.append(result['PRF']['Precision'][3])
#     Recall2.append(result['PRF']['Recall'][2])
#     Recall3.append(result['PRF']['Recall'][3])

# import matplotlib.pyplot as plt
# #
# #fig, ax = plt.subplots()
# ##
# ##ax.plot(lambdas, Precision2, label='Precision2')
# ##ax.plot(lambdas, Precision3, label='Precision3')
# ##ax.plot(lambdas, Recall2, label='Recall2')
# ##ax.plot(lambdas, Recall3, label='Recall3')
# ##
# ##ax.legend()
# #
# fig, ax = plt.subplots()

# ax.plot(betas, Precision2, label='Precision2')
# ax.plot(betas, Precision3, label='Precision3')
# ax.plot(betas, Recall2, label='Recall2')
# ax.plot(betas, Recall3, label='Recall3')

# ax.legend()
# #
